docs/designs/CodeFlare-SDK-design-doc.md

```diff
@@ -0,0 +1,137 @@
+
+On a high level, the interaction between a data scientist, CodeFlare SDK and the rest of the components are explained as below:
+
+![Diagram-1](../images/codeflare_stack_arch.png)
+
+The role of the SDK in the model training and tuning stack remains consistent on the client side, regardless of any changes to the components it interacts with on the cluster.
+
+![Diagram-2](../images/codeflare_sdk.png)
+
+#### Considerations while designing SDK:
+1. Ease of use and integration: The SDK’s primary role is to abstract Kubernetes specifics. It should provide simple interfaces for interacting with any of the model training components on the server side.
+2. Lightweight: The SDK runs client-side and should minimize resource consumption. It must prioritize responsiveness and user experience. For example, using a polling mechanism to fetch status instead of a constant resource watch.
+3. Extensibility: The SDK currently integrates well with the CodeFlare stack, which uses Ray and TorchX (pytorch) distributed framework. In the future, components used for distributed training/tuning (as seen in fig [2]) should remain interchangeable.
+4. Security: The SDK must ensure users see only the information they are authorized to view. Authentication occurs on the client, while admission and validation are handled server-side. Kubernetes RBAC makes things easier.
+5. Typed Object Creation: The client should only allow the creation of known, typed K8s resources. This prevents arbitrary payloads from reaching the server which could be a threat. (configuring AppWrapper resource template is a concern here).
+6. Version Compatibility: The SDK must maintain compatibility between client and server versions. Backward compatibility should be ensured even if one side is upgraded.
+
+#### Codebase Modularization:
+The CodeFlare-SDK codebase requires refactoring and modularization to facilitate easier addition or modification of components. The key requirements for this refactor include:
+
+1. Component Structure: The SDK will be divided into distinct components:
+  - Modules for commands, helpers, and utilities that interact with Kubernetes (OpenShift) clusters.
+  - Helpers specific to framework clusters like Ray.
+2. Self-Contained Packages: Each Python package will function as an independent, self-contained module with its own dedicated functionality. This will make it easier to manage, maintain, and extend the SDK.
+3. Testing: Unit tests for each API will be located within their respective submodules to ensure:
+  - Comprehensive test coverage, closely tied to the implementation, facilitating easier association between code and tests.
+  - Integration and upgrade tests that span multiple modules will be managed separately, ensuring robust validation of both individual and end-to-end functionality.
```

Comment:
I have created a separate section for code coverage, and mentioned codecov as an example for external tools so that later if we find something better than that we can adopt it.